### PR TITLE
Assignment6-Cynthia and Hadi

### PR DESCRIPTION
**Pull request by HadiM2**
@HadiM2 @montalcy

### CODE REVIEW

**Comment by pja6** [comment]
Will review

File: Calculate/Program.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+        do
+        {
+            Pro.WriteLine("Enter your expression: ");
+            input = Pro.ReadLine()!;
+            if (calc.TryCalculate(input, out solution))
+            {
+                Pro.WriteLine($"Answer: {solution}");
+            }
+        } while (input == null || !calc.TryCalculate(input, out solution));
```
Comment (pja6) [inline Calculate/Program.cs:31]: This looks great!

File: Calculate/Calculator.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    public IReadOnlyDictionary<char, Func<int, int, int>> MathematicalOperations { get; } = new Dictionary<char, Func<int, int, int>>
+    {
+        { '+', Add},
+        { '-' , Subtract },
+        { '/', Divide },
+        {'*', Multiply }
+    };
+
+    public static int Add(int firstValue, int secondValue) => firstValue + secondValue;
+
```
Comment (pja6) [inline Calculate/Calculator.cs:19]: Nice job using the lambda expressions

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    public static int Subtract(int firstValue, int secondValue) => firstValue - secondValue;
+    public static int Multiply(int firstValue, int secondValue) => (firstValue * secondValue);
+    public static int Divide(int firstValue, int secondValue) => secondValue != 0 ? (firstValue / secondValue) : throw new ArgumentException("Cant divide by zero");
```
Comment (pja6) [inline Calculate/Calculator.cs:22]: Great job checking for this

File: Calculate.Tests/CalculatorTest.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        int solution = Calculator.Multiply(2, 2);
+        Assert.Equal(4, solution);
+    }
+
+    [Fact]
+    public void DivideByZeroFails()
+    {
+        Assert.Throws<ArgumentException>(()=> Calculator.Divide(2, 0));
+    }
+
```
Comment (pja6) [inline Calculate.Tests/CalculatorTest.cs:46]: Looks good

**Review by pja6 — COMMENTED** [review]
Instructions:
* Create a Console project called "Calculate.". ✔
* Define a Program Class
    * Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
    * Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌// I didn’t see any tests for program
    * Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔

* Define a Calculator class ✔
    * Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
    * Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
        * is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
            * Uses char for the key corresponding to the operators +, -, *, and /. ✔
            * Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
    * Implement a TryCalculate method following "TryParse" pattern ✔
        * Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
        * If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
        * Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
        * Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
* Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
* Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌ //would this be in the program tests?


Fundamentals:
* Place all shared project properties into a Directory.Build.Props file. ✔
* Place all shared project items into a Directory.Build.targets file.
* nullable reference types is enabled ✔
* Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
* Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
* and enabled .NET analyzers for both projects ✔
* For this assignment, always use Assert.AreEqual<T>() (the generic version) ✔
* All of the above should be unit tested ❌ // Missing program tests
* Choose simplicity over complexity ✔

Everything looks really good and matches specifications for the most part - I think you might be missing tests for program though, other than that nothing else stood out for correction - Nice job!

**Comment by MattA41** [comment]
Im taking a look at this

File: Directory.Build.Props [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+	<PropertyGroup>
+		<TargetFramework>net8.0</TargetFramework>
+		<ImplicitUsings>enable</ImplicitUsings>
+		<Nullable>enable</Nullable>
+		<LangVersion>12.0</LangVersion>
+		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+		<EnableNETAnalyzers>true</EnableNETAnalyzers>
+		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
+		<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+	</PropertyGroup>
```
Comment (MattA41) [inline Directory.Build.Props:11]: nice build props file

File: Calculate/Calculator.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    public bool TryCalculate(string expression, out int solution)
+    {
+        solution = 0;
+        string[] inputs= expression.Split(' ');
+        if (inputs.Length != 3) return false;
+        if (!int.TryParse(inputs[0], out int firstint)) return false;
+        if (!int.TryParse(inputs[1], out int secondint)) return false;
```
Comment (MattA41) [inline Calculate/Calculator.cs:30]: If I am reading this right you are making a string array that splits the strings on the spaces. That would look something like this "3 + 4", inputs[0] = 3, inputs[1] = +, inputs[2] = 4. It looks like you are calling TryParse on the operand in the solution

**Review by MattA41 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
    Define a Program Class
        Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
        Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ❌
        Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
    Define a Calculator class ✔
        Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
        Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
            is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
                Uses char for the key corresponding to the operators +, -, *, and /. ✔
                Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
        Implement a TryCalculate method following "TryParse" pattern ✔
            Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ❌✔
            If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
            Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
            Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
    Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
    Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌

Extra Credit

Do one of the following two options (or both if you want extra, extra credit) :)

    Refactor the redirect portion of the Program class into 'ProgramBase`
    Move ProgramBase into a ConsoleUtilities assembly to be used in other console-based projects
    Use generics the mathematical operations methods and consider using generic constraints (requires .NET 7.0)

Fundamentals

    Place all shared project properties into a Directory.Build.Props file.
    Place all shared project items into a Directory.Build.targets file.
    nullable reference types is enabled ✔
    Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    and enabled .NET analyzers for both projects ✔
    For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
    All of the above should be unit tested ❌
    Choose simplicity over complexity ✔

File: Calculate/Calculator.cs (hunk 6 above)
Comment (montalcy) [reply Calculate/Calculator.cs:30]: Thank you for catching this! Was just failing my test because of this miss.

File: Calculate/Program.cs (hunk 1 above)
Comment (montalcy) [reply Calculate/Program.cs:33]: Thanks!

File: Calculate/Calculator.cs (hunk 2 above)
Comment (montalcy) [reply Calculate/Calculator.cs:19]: Thanks! They really do come in handy

File: Calculate.Tests/CalculatorTest.cs (hunk 4 above)
Comment (HadiM2) [reply Calculate.Tests/CalculatorTest.cs:104]: Thank you.

File: Directory.Build.Props (hunk 5 above)
Comment (HadiM2) [reply Directory.Build.Props:11]: Thank you Matt : )

File: Calculate/Calculator.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        { '-' , Subtract },
+        { '/', Divide },
+        {'*', Multiply }
+    };
+
+    public static int Add(int firstValue, int secondValue) => firstValue + secondValue;
+
+    public static int Subtract(int firstValue, int secondValue) => firstValue - secondValue;
+    public static int Multiply(int firstValue, int secondValue) => (firstValue * secondValue);
+    public static int Divide(int firstValue, int secondValue) => secondValue != 0 ? (firstValue / secondValue) : throw new ArgumentException("Cant divide by zero");
```
Comment (BenjaminMichaelis) [inline Calculate/Calculator.cs:22]: Divide by zero by default throws a `System.DivideByZeroException`

**Pull request closed without merging**